airbyte-integrations/connectors/source-bing-ads/source_bing_ads/source.py

```diff
@@ -0,0 +1,325 @@
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+
+
+from abc import ABC, abstractmethod
+from typing import Any, Iterable, List, Mapping, MutableMapping, Optional, Tuple
+
+from airbyte_cdk import AirbyteLogger
+from airbyte_cdk.models import SyncMode
+from airbyte_cdk.sources import AbstractSource
+from airbyte_cdk.sources.streams import Stream
+from source_bing_ads.cache import VcrCache
+from source_bing_ads.client import Client
+from suds import sudsobject
+
+CACHE: VcrCache = VcrCache()
+
+
+class BingAdsStream(Stream, ABC):
+    primary_key = "Id"
+
+    def __init__(self, client: Client, config: Mapping[str, Any]) -> None:
+        super().__init__()
+        self.client = client
+        self.config = config
+
+    @property
+    @abstractmethod
+    def data_field(self) -> str:
+        """
+        Specifies root object name in a stream response
+        """
+        pass
+
+    @property
+    @abstractmethod
+    def service_name(self) -> str:
+        """
+        Specifies bing ads service name for a current stream
+        """
+        pass
+
+    @property
+    @abstractmethod
+    def operation_name(self) -> str:
+        """
+        Specifies operation name to use for a current stream
+        """
+        pass
+
+    @property
+    @abstractmethod
+    def additional_fields(self) -> Optional[str]:
+        """
+        Specifies which additional fields to fetch for a current stream.
+        Expected format: field names separated by space
+        """
+        pass
+
+    def next_page_token(self, response: sudsobject.Object, **kwargs: Mapping[str, Any]) -> Optional[Mapping[str, Any]]:
+        """
+        Default method for streams that don't support pagination
+        """
+        return None
+
+    def parse_response(self, response: sudsobject.Object, **kwargs) -> Iterable[Mapping]:
+        if response is not None and hasattr(response, self.data_field):
+            yield from self.client.asdict(response)[self.data_field]
+
+        yield from []
+
+    def send_request(self, params: Mapping[str, Any], account_id: str = None) -> Mapping[str, Any]:
+        with CACHE.use_cassette():
```

Comment:
we should not cache all possible requests. - otherwise, we could run out of memory.

We only need to cache requests/streams which are called a lot of times. 

Btw - could you please add comments in code and in review  with list of streams/requests which requires cache

Reply:
Ads stream doesn't use cache, other 3 use it. And I'm fetching accounts stream on each stream

Reply:
added `use_cache` field to disable/enable caching for stream